torch/utils/mkldnn.py

```diff
@@ -0,0 +1,34 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import torch
+
+
+def to_mkldnn(module):
+    for child in module.children():
+        to_mkldnn(child)
+
+    def fn(t):
+        if t.is_floating_point():
+            return t.to_mkldnn()
+
+    for param in module._parameters.values():
+        if param is not None:
+            # Tensors stored in modules are graph leaves, and we don't
+            # want to create copy nodes, so we have to unpack the data.
+            param.data = fn(param.data)
+            if param._grad is not None:
+                param._grad.data = fn(param._grad.data)
+
+    for key, buf in module._buffers.items():
+        if buf is not None:
+            module._buffers[key] = fn(buf)
+
+    if isinstance(module, torch.nn.Conv2d):
+        module.weight.data = torch._C._nn.mkldnn_reorder_conv2d_weight(
```

Comment:
@mingfeima - so in general - is this reordering for inference only? if we do it in training, we'd need to make sure that all optimizers can handle different layout/dimensions and even auxiliary weights like for Adam work well.

Reply:
can we just do the reorder on the fly in the function?  I'm unclear if the reorder is expensive -- it looks pretty simple, but not sure how it's actually implemented.

Reply:
@dzhulgakov Pre-reordering weights is still useful for training even though perf gain might not be that obvious, depending on the workloads. It is possible to leave existing optimizers intact to support the opaque tensor as long as it supports all the ops and factory functions used by the optimizers.

Reply:
reorder on the fly (repeatedly) will cause 20% slower in conv (and 15% slower in the whole resnext).

Reply:
is that slower than non-mkldnn or slower than mkldnn with optimization?

Reply:
What I'm trying to get at: if the "on the fly" packing is still faster than the dense path, then it makes sense to expose this op independent of pre-packing.  But if it's not, we need to tie this into our greater weight pre-packing story.